### ISSUE TITLE
IngressRoute: Skip service if an error occurs

### ISSUE THREAD

**Issue by ErikLundJensen** [traefik/traefik#9578]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

Kubernetes CRD objects IngressRoute and IngressRouteTCP have different error handling semantic.
Given IngressRoute with a route that contains multiple services then one failing service will break the chain.
For example if there are no healthy pods behind one of the services then `404` "Not Found" is returned - even if another service has healthy pods.

IngressRouteTCP does not have this behaviour as it sends the traffic to the healthy pods and thereby not breaking the chain.

We suggest that the semantic of IngressRoute and IngressRouteTCP should be the same.

One of the benefit from this is that Traefik can send traffic to another cluster in case a service becomes unavailable in the local cluster. An example of IngressRoute with two services is shown below.

```yml
kind: IngressRoute
apiVersion: traefik.containo.us/v1alpha1
metadata:
  name: hello-world-with-two-services
spec:
  entryPoints:
    - websecure
  routes:
    - kind: Rule
      match: Host(`hello-world.apps.foo.bar`)
      services:
        - kind: Service
          name: unhealthy-service
          namespace: demo
          port: 80
        - kind: Service
          name: healthy-service
          namespace: demo
          port: 80    
  tls:
    secretName: hello-world-tls
```

See also PR #9547 for a simple implementation. This feature request replaces bug issue #9547 

This change could be a candidate for [the epic tracking CRD improvements](https://github.com/traefik/traefik/issues/9125)

**Comment by ErikLundJensen** [comment]
Multi-cluster traffic routing in Traefik is a challinging task (ref [Traefik Lab Blog](https://traefik.io/blog/key-challenges-of-multi-cluster-networking/) 
However, this can be simplified by using external services in the routes as described above. The implementation already is inplace for IngressRouteTCP routes, but fails for IngressRoute routes.

For example by creating an IngressRoute with reference to an external service `hello-world-west`. This configuration will keep traffic local inside the cluster as long as the service is backed by healthy pods. The extra header match is added to avoid recursive calls if the service is not available in any of the clusters.

```
kind: IngressRoute
apiVersion: traefik.containo.us/v1alpha1
metadata:
  annotations:
    kubernetes.io/ingress.class: traefik
  name: hello-world-with-failover
spec:
  entryPoints:
    - websecure
  routes:
    - kind: Rule
      match: "Host(`hello-world.foo.bar`) && !Headers(`X-FOO-BAR-TRAEFIK-FAILOVER`, `true`)"
      priority: 1
      middlewares:
        - name: failover-header
      services:
        - kind: Service
          name: hello-world
          port: 80
          weight: 10000000
        - kind: Service
          name: hello-world-west
          port: 443
          scheme: https
          weight: 1
    # No failover if request already comes from Traefik in other site
    - kind: Rule
      match: "Host(`hello-world.foo.bar`) && Headers(`X-FOO-BAR-TRAEFIK-FAILOVER`, `true`)"
      priority: 10
      services:
        - kind: Service
          name: hello-world
          port: 80
  tls:
    secretName: hello-world-tls
```

**Comment by ErikLundJensen** [comment]
@nmengin Please, prioritize this proposal as it simplifies implementation failover between two cluster. Traefik's TCP implementation already supports this and I would say that it is a bug that Traefik's HTTP implementation does not have same fault handling.

**Comment by GiuseppeChiesa-TomTom** [comment]
We also would like to leverage this pattern to allow failover on another cluster/external-service (as per code snippet above). If it can get prioritized it would be very nice feature for multiregion deployments.

**Comment by tfny** [comment]
Heya, we are interested in this, but we need to take some time to look more deeply at it as we think it might bring breaking changes.  There are also many ways to tackle the modifications if we decide to do it, so that is going to take some time. 

If you are interested in helping, please feel free to share your ideas and thoughts on the subject.

**Comment by ErikLundJensen** [comment]
Could this be part of the upcoming milestone-3.0 or has this request prioritized down?

**Comment by ErikLundJensen** [comment]
Any news? I really hope this can be part of the 3.0 release.

**Comment by agmimidi** [comment]
Hey all,

Will this be part of the final 3.0 release?

**Comment by ErikLundJensen** [comment]
@juliens Any statement from you regarding getting this fix into the 3.0 release?

**Comment by juliens** [comment]
@ErikLundJensen 

The issue is currently in the proposal stage, making it challenging to incorporate into v3 since the release candidate is expected imminently.

From my point of view, the problem isn't very clear. I wonder if there are other features that would meet your needs. Consider exploring the following features:
- https://doc.traefik.io/traefik/providers/kubernetes-crd/#allowemptyservices
- https://doc.traefik.io/traefik/routing/services/#failover-service

**Comment by ErikLundJensen** [comment]
Thanks for replying.
The AllowEmptyService only gives another error code, it does not change the behaviour of service invocations. The link to the failover-service describes a setup that is not supported by the Kubernetes provider.

I have tested the [TraefikService weighted-round-robin](https://doc.traefik.io/traefik/routing/providers/kubernetes-crd/#weighted-round-robin). The weighted-round-robin fails in the timeframe from the pod starts to it becomes ready. I need to debug more to see why the requests are sent to a Kubernetes service that does not have any pods in ready-state.


Anyhow, I can try to explain the problem with IngressRoute in other words. When you have a route with several services then we see different behaviour in the IngressRoute (http) and the IngressRouteTCP in Kubernetes.

Given service A is unavailable and service B is available
and a rule has defined with two services A and B
then the IngressRouteTCP continues to the next service B
but the IngressRoute (http) returns error.
 
The semantic error handling in IngressRouteTCP and IngressRoute (http) should be the same.

For example the IngressRoute below will return 404 if the service named "unhealthy-service" is unavailable.
However, if the IngressRoute is changed to a IngressRouteTCP then the service named "healthy-service" is invoked.

```
kind: IngressRoute
apiVersion: traefik.containo.us/v1alpha1
metadata:
  name: hello-world-with-two-services
spec:
  entryPoints:
    - websecure
  routes:
    - kind: Rule
      match: Host(`hello-world.apps.foo.bar`)
      services:
        - kind: Service
          name: unhealthy-service
          namespace: demo
          port: 80
        - kind: Service
          name: healthy-service
          namespace: demo
          port: 80    
  tls:
    secretName: hello-world-tls
```

We do see this as an error in the implementation of error handling.
The use-cases of having multiple services in a rule are many. One of them is handling fail-over.

I you prefer I can create a new issue related to the TraefikService if you prefer the fail-over to be handled by TraefikService instead.

**Comment by juliens** [comment]
Thanks for your answer,

There is already an issue related to TraefikService and failover https://github.com/traefik/traefik/issues/9919, if you agree that this is the same proposal, we can keep only one.

**Comment by ErikLundJensen** [comment]
Closed as will be fixed by #9919
